src/serac/physics/integrators/displacement_hyperelastic_integrator.cpp

```diff
@@ -0,0 +1,217 @@
+#include "serac/numerics/expr_template_ops.hpp"
+#include "serac/numerics/array_4D.hpp"
+#include "serac/physics/utilities/physics_utils.hpp"
+
+namespace serac::mfem_ext {
+
+void DisplacementHyperelasticIntegrator::CalcKinematics(const mfem::FiniteElement&    element,
+                                                        const mfem::IntegrationPoint& int_point,
+                                                        mfem::ElementTransformation& parent_to_reference_transformation)
+{
+  // Calculate the reference to stress-free transformation
+  CalcInverse(parent_to_reference_transformation.Jacobian(), dxi_dX_);
+
+  // Calculate the derivatives of the shape functions in the reference space
+  element.CalcDShape(int_point, dN_dxi_);
+
+  // Calculate the derivatives of the shape functions in the stress free configuration
+  Mult(dN_dxi_, dxi_dX_, dN_dX_);
+
+  // Calculate the displacement gradient using the current DOFs
+  MultAtB(input_state_matrix_, dN_dX_, du_dX_);
+
+  solid_util::calcDeformationGradient(du_dX_, F_);
+
+  // Calculate the inverse of the deformation gradient
+  mfem::CalcInverse(F_, Finv_);
+
+  // Calculate the B matrix (dN/Dx where x is the current configuration)
+
+  if (geom_nonlin_) {
+    // If we're including geometric nonlinearities, we integrate on the current deformed configuration
+    mfem::Mult(dN_dX_, Finv_, B_);
+    det_J_ = F_.Det();
+  } else {
+    // If not, we integrate on the undeformed stress-free configuration
+    B_     = dN_dX_;
+    det_J_ = 1.0;
+  }
+}
+
+double DisplacementHyperelasticIntegrator::GetElementEnergy(
+    const mfem::FiniteElement& element, mfem::ElementTransformation& parent_to_reference_transformation,
+    const mfem::Vector& state_vector)
+{
+  int dof = element.GetDof();
+  int dim = element.GetDim();
+
+  // Reshape the state vector
+  input_state_matrix_.UseExternalData(state_vector.GetData(), dof, dim);
+
+  // Determine the integration rule from the order of the FE space
+  const mfem::IntegrationRule* ir = IntRule;
+  if (!ir) {
+    ir = &(mfem::IntRules.Get(element.GetGeomType(), 2 * element.GetOrder() + 3));
+  }
+
+  double energy = 0.0;
+
+  // Set the transformation for the underlying material. This is required for coefficient evaluation.
+  material_.setTransformation(parent_to_reference_transformation);
+
+  for (int i = 0; i < ir->GetNPoints(); i++) {
+    // Set the current integration point
+    const mfem::IntegrationPoint& int_point = ir->IntPoint(i);
+    parent_to_reference_transformation.SetIntPoint(&int_point);
+
+    // Calculate the deformation gradent and accumulate the strain energy at the current integration point
+    CalcKinematics(element, int_point, parent_to_reference_transformation);
+    energy += det_J_ * int_point.weight * parent_to_reference_transformation.Weight() * material_.evalStrainEnergy(F_);
+  }
+
+  return energy;
+}
+
+void DisplacementHyperelasticIntegrator::AssembleElementVector(
```

Comment:
I guess if you had a mask, there are a number of levels you could operate the mask at, so one way would just to "run two top-level integrators", in which case my comments about turning "DisplacementHyperelasticIntegrator::CalcKinematics" into "displacementsolidmechanicsintegrator::CalcHyperelasticKinematics" might not be relevant.  Again probably OBE by whatever Sam comes up with

Reply:
Agreed. That's probably how we would do it if we didn't have the incoming weak form rewrite.